src/dbnode/client/session.go

```diff
@@ -2322,7 +2332,226 @@ func (s *session) streamBlocksMetadataFromPeer(
+		return s.streamBlocksMetadataFromPeerSequential(namespace, shard, peer,
+			start, end, startPageToken, metadataCh, resultOpts, progress, nil)
+	}
+
+	var (
+		optionIncludeSizes     = true
+		optionIncludeChecksums = true
+		optionIncludeLastRead  = true
+	)
+
+	// Get first page with size 1 in flush phase.
+	token := &pagetoken.PageToken{
+		FlushedSeriesPhase: &pagetoken.PageToken_FlushedSeriesPhase{},
+	}
+	requestToken, err := proto.Marshal(token)
+	if err != nil {
+		return nil, err
+	}
+
+	// First start at zero and retrieve the volume index.
+	tctx, _ := thrift.NewContext(s.streamBlocksMetadataBatchTimeout)
+	req := rpc.NewFetchBlocksMetadataRawV2Request()
+	req.NameSpace = namespace.Bytes()
+	req.Shard = int32(shard)
+	req.RangeStart = start.UnixNano()
+	req.RangeEnd = end.UnixNano()
+	req.Limit = 1
+	req.PageToken = requestToken
+	req.IncludeSizes = &optionIncludeSizes
+	req.IncludeChecksums = &optionIncludeChecksums
+	req.IncludeLastRead = &optionIncludeLastRead
+
+	var (
+		result *rpc.FetchBlocksMetadataRawV2Result_
+	)
+	borrowErr := peer.BorrowConnection(func(client rpc.TChanNode) {
+		result, err = client.FetchBlocksMetadataRawV2(tctx, req)
+	})
+	if borrowErr != nil {
+		return nil, borrowErr
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if result.NextPageToken == nil {
+		// Complete, nothing for this time range.
+		return nil, nil
+	}
+
+	// Decode page token.
+	token = &pagetoken.PageToken{}
+	if err := proto.Unmarshal(result.NextPageToken, token); err != nil {
+		return nil, err
+	}
+
+	// Now get the volume.
+	if token.ActiveSeriesPhase != nil || token.FlushedSeriesPhase == nil {
+		// This is an error case.
+		return nil, fmt.Errorf("flushed phase not set for block start: %v", start)
+	}
+	if token.FlushedSeriesPhase.CurrBlockStartUnixNanos != start.UnixNano() {
+		// This is an error case.
+		return nil, fmt.Errorf("flushed phase zero value for block start: %v", start)
+	}
+
+	// We can directly take value of it.
+	volume := token.FlushedSeriesPhase.Volume
+
+	var (
+		prevProbePos int64
+		probePos     int64
+		wg           sync.WaitGroup
+		errors       xerrors.MultiError
+		errorsLock   sync.Mutex
+	)
+	streamSequentialWorkers := xsync.NewWorkerPool(
+		s.opts.(AdminOptions).FetchSeriesBlocksBatchConcurrency())
+	streamSequentialWorkers.Init()
+
+	enqueueStream := func() error {
+		startPos := prevProbePos
+		limitTotal := probePos - prevProbePos
+
+		wg.Add(1)
+		streamSequentialWorkers.Go(func() {
+			defer wg.Done()
+
+			// TODO: cancel inflight metadata calls if we return early
+			// from an error in the dispatch loop.
+			_, err := s.streamBlocksMetadataFromPeerSequential(namespace, shard, peer,
+				start, end, nil, metadataCh, resultOpts, progress,
+				&streamBlocksMetadataFromPeerSequentialOptions{
+					startPos:   startPos,
+					limitTotal: limitTotal,
+					volume:     volume,
+				})
+			if err != nil {
+				errorsLock.Lock()
+				errors = errors.Add(err)
+				errorsLock.Unlock()
```

Comment:
we should return error here, no?

Reply:
We check this at the end, can't return error here unfortunately since it's an async goroutine.